libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +684,430 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
-                        __memcpy_async_sm_dispatch<_Native_alignment>(__group, __destination, __source, __size, __sync, __is_async);
-                    case 4: __is_async = __memcpy_async_impl<8>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                    case 3: __is_async = __memcpy_async_impl<4>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                    case 2: // fallthrough
-                    case 1: __is_async = __memcpy_async_impl<1>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                }
-            }
-            else {
-                __is_async = __memcpy_async_impl<_Native_alignment>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size());
+template <>
+inline __device__
+void __cp_async_shared_global<16>(char * __dest, const char * __src) {
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async
+    // When copying 16 bytes, it is possible to skip L1 cache (.cg).
+    asm volatile(
+        "cp.async.cg.shared.global [%0], [%1], %2, %2;"
+        :
+        : "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__dest))),
+          "l"(static_cast<_CUDA_VSTD::uint64_t>(__cvta_generic_to_global(__src))),
+          "n"(16)
+        : "memory");
+}
+
+template <size_t _Alignment, typename _Group>
+inline __device__
+void __cp_async_shared_global_mechanism(_Group __g, char * __dest, const char * __src, _CUDA_VSTD::size_t __size) {
+    static_assert(4 <= _Alignment, "cp.async requires at least 4-byte alignment");
+
+    // Maximal copy size is 16.
+    constexpr int __copy_size = (_Alignment > 16) ? 16 : _Alignment;
+    // We use an int offset here, because we are copying to shared memory,
+    // which is easily addressable using int.
+    const int __group_size = __g.size();
+    const int __group_rank = __g.thread_rank();
+    const int __stride = __group_size * __copy_size;
+    for (int __offset = __group_rank *  __copy_size; __offset < __size; __offset += __stride) {
+        __cp_async_shared_global<__copy_size>(__dest + __offset, __src + __offset);
+    }
+}
+#endif // __CUDA_MINIMUM_ARCH__
+
+template <size_t _Copy_size>
+struct __copy_chunk {
+    _ALIGNAS(_Copy_size) char data[_Copy_size];
+};
+
+template <size_t _Alignment, typename _Group>
+inline __host__ __device__
+void __cp_async_fallback_mechanism(_Group __g, char * __dest, const char * __src, _CUDA_VSTD::size_t __size) {
+    // Maximal copy size is 16 bytes
+    constexpr _CUDA_VSTD::size_t __copy_size = (_Alignment > 16) ? 16 : _Alignment;
+    using __chunk_t = __copy_chunk<__copy_size>;
+
+    // "Group"-strided loop over memory
+    const size_t __stride = __g.size() * __copy_size;
+    for (_CUDA_VSTD::size_t __offset = __g.thread_rank() * __copy_size; __offset < __size; __offset += __stride) {
```

Comment:
This loop needs a pragma unroll 64 to limit the amount of code generated for degenerate cases. However, this results in errors about unknown pragmas in one of the GCC compilation phases. 

I wanted to add this to `__config`: 
``` cuda
#if defined(_LIBCUDACXX_CUDACC)
#  define _LIBCUDACXX_PRAGMA_UNROLL(_N) _Pragma(_LIBCUDACXX_TOSTRING(unroll _N))
#else
#  define _LIBCUDACXX_PRAGMA_UNROLL(_N)
#endif // _LIBCUDACXX_COMPILER_CUDACC
```
This doesn't work however. @miscco : do you know how to get this to work?

This is the code I removed: 
``` cuda
    // An unroll factor of 64 ought to be enough for anybody. This unroll pragma
    // is mainly intended to place an upper bound on loop unrolling. The number
    // is more than high enough for the intended use case: an unroll factor of
    // 64 allows moving 4 * 64 * 256 = 64kb in one unrolled loop with 256
    // threads (copying ints). On the other hand, in the unfortunate case that
    // we have to move 1024 bytes / thread with char width, then we prevent
    // fully unrolling the loop to 1024 copy instructions. This prevents the
    // compile times from increasing unreasonably, and also has neglibible
    // impact on runtime performance.
_LIBCUDACXX_PRAGMA_UNROLL(64)
```

Reply:
Yes, you need to change this to account for msvc
```cpp
#if defined(_LIBCUDACXX_CUDACC)
#if defined(_LIBCUDACXX_COMPILER_MSVC
#  define _LIBCUDACXX_PRAGMA_UNROLL(_N) __pragma(_LIBCUDACXX_TOSTRING(unroll _N))
#else // ^^^ _LIBCUDACXX_COMPILER_MSVC ^^^ / vvv !_LIBCUDACXX_COMPILER_MSVC vvv
#  define _LIBCUDACXX_PRAGMA_UNROLL(_N) _Pragma(_LIBCUDACXX_TOSTRING(unroll _N))
#endif // !_LIBCUDACXX_COMPILER_MSVC
#else // ^^^ _LIBCUDACXX_CUDACC ^^^ / vvv !_LIBCUDACXX_CUDACCvvv
#  define _LIBCUDACXX_PRAGMA_UNROLL(_N)
#endif // !_LIBCUDACXX_COMPILER_CUDACC
```